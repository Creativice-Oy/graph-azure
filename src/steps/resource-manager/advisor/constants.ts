import { RelationshipClass } from '@jupiterone/integration-sdk-core';
import { ANY_SCOPE } from '../constants';

export const AdvisorSteps = {
  RECOMMENDATIONS: 'rm-advisor-recommendations',
};

export const AdvisorEntities = {
  RECOMMENDATION: {
    _type: 'azure_advisor_recommendation',
    _class: ['Finding'],
    resourceName: '[RM] Advisor Recommendation',
  },
};

export const AdvisorRelationships = {
  ANY_RESOURCE_HAS_FINDING: {
    _type: 'ANY_SCOPE_has_finding',
    sourceType: ANY_SCOPE,
    _class: RelationshipClass.HAS,
    targetType: AdvisorEntities.RECOMMENDATION._type,
  },
};
