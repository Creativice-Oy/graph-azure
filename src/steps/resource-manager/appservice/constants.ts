import { createResourceGroupResourceRelationshipMetadata } from '../utils/createResourceGroupResourceRelationship';

export const AppServiceSteps = {
  APPS: 'rm-appservice-apps',
};

export const AppServiceEntities = {
  WEB_APP: {
    _type: 'azure_web_app',
    _class: ['Application'],
    resourceName: '[RM] Web App',
  },
  FUNCTION_APP: {
    _type: 'azure_function_app',
    _class: ['Function'],
    resourceName: '[RM] Function App',
  },
};

export const AppServiceRelationships = {
  RESOURCE_GROUP_HAS_WEB_APP: createResourceGroupResourceRelationshipMetadata(
    AppServiceEntities.WEB_APP._type,
  ),
  RESOURCE_GROUP_HAS_FUNCTION_APP: createResourceGroupResourceRelationshipMetadata(
    AppServiceEntities.FUNCTION_APP._type,
  ),
};
