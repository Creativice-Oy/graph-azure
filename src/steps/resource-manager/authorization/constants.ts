import { PrincipalType } from '@azure/arm-authorization/esm/models';
import {
  GROUP_ENTITY_TYPE,
  STEP_AD_GROUPS,
  USER_ENTITY_TYPE,
  STEP_AD_USERS,
  SERVICE_PRINCIPAL_ENTITY_TYPE,
  STEP_AD_SERVICE_PRINCIPALS,
} from '../../active-directory';
import { generateRelationshipType } from '@jupiterone/integration-sdk-core';

// Fetch Role Assignments & Definitions
export const STEP_RM_AUTHORIZATION_ROLE_ASSIGNMENTS_AND_DEFINITIONS =
  'rm-authorization-role-assignments-and-definitions';

export const ROLE_DEFINITION_ENTITY_TYPE = 'azure_role_definition';
export const ROLE_DEFINITION_ENTITY_CLASS = ['AccessRole', 'AccessPolicy'];

export const ROLE_ASSIGNMENT_RELATIONSHIP_CLASS = 'ASSIGNED';

interface RoleAssignmentMap {
  principalType: PrincipalType;
  type: string;
  dependsOn: string[];
}

export const ROLE_ASSIGNMENT_TYPES_MAP: RoleAssignmentMap[] = [
  {
    principalType: 'Application',
    type: 'azure_application',
    dependsOn: [],
  },
  {
    principalType: 'DirectoryObjectOrGroup',
    type: 'azure_directory',
    dependsOn: [],
  },
  {
    principalType: 'DirectoryRoleTemplate',
    type: 'azure_directory_role_template',
    dependsOn: [],
  },
  {
    principalType: 'Everyone',
    type: 'azure_everyone',
    dependsOn: [],
  },
  {
    principalType: 'ForeignGroup',
    type: 'azure_foreign_group',
    dependsOn: [],
  },
  {
    principalType: 'Group',
    type: GROUP_ENTITY_TYPE,
    dependsOn: [STEP_AD_GROUPS],
  },
  {
    principalType: 'MSI',
    type: 'azure_msi',
    dependsOn: [],
  },
  {
    principalType: 'ServicePrincipal',
    type: SERVICE_PRINCIPAL_ENTITY_TYPE,
    dependsOn: [STEP_AD_SERVICE_PRINCIPALS],
  },
  {
    principalType: 'Unknown',
    type: 'azure_unknown',
    dependsOn: [],
  },
  {
    principalType: 'User',
    type: USER_ENTITY_TYPE,
    dependsOn: [STEP_AD_USERS],
  },
];

export const ROLE_ASSIGNMENT_DEFAULT_TYPE = 'azure_unknown_role_target';

export const ROLE_ASSIGNMENT_DEPENDS_ON = ([] as string[]).concat(
  ...ROLE_ASSIGNMENT_TYPES_MAP.map((t) => t.dependsOn),
);
export const ROLE_ASSIGNMENT_TARGET_ENTITY_TYPES = ([] as string[]).concat(
  ...ROLE_ASSIGNMENT_TYPES_MAP.map((t) => t.type),
);

export function getJupiterTypeForPrincipalType(
  principalType: PrincipalType | undefined,
): string {
  return (
    ROLE_ASSIGNMENT_TYPES_MAP.find((t) => t.principalType === principalType)
      ?.type || ROLE_ASSIGNMENT_DEFAULT_TYPE
  );
}

export function createRoleAssignmentRelationshipType(
  targetEntityType: string,
): string {
  return generateRelationshipType(
    ROLE_ASSIGNMENT_RELATIONSHIP_CLASS,
    ROLE_DEFINITION_ENTITY_TYPE,
    targetEntityType,
  );
}

export const ROLE_ASSIGNMENT_RELATIONSHIP_TYPES = [
  createRoleAssignmentRelationshipType(ROLE_ASSIGNMENT_DEFAULT_TYPE),
  ...ROLE_ASSIGNMENT_TARGET_ENTITY_TYPES.map(
    createRoleAssignmentRelationshipType,
  ),
];

// Fetch Classic Administrators
export const STEP_RM_AUTHORIZATION_CLASSIC_ADMINISTRATORS =
  'rm-authorization-classic-administrators';

export const CLASSIC_ADMINISTRATOR_ENTITY_KEY = 'azure_classic_admin_group';
export const CLASSIC_ADMINISTRATOR_ENTITY_TYPE = 'azure_classic_admin_group';
export const CLASSIC_ADMINISTRATOR_RELATIONSHIP_TYPE =
  'azure_classic_admin_group_has_user';
export const CLASSIC_ADMINISTRATOR_ENTITY_CLASS = 'UserGroup';
