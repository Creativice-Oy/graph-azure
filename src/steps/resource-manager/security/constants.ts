import { RelationshipClass } from '@jupiterone/integration-sdk-core';
import { SUBSCRIPTION_ENTITY_METADATA } from '../subscriptions';

export const SecuritySteps = {
  ASSESSMENTS: 'rm-security-assessments',
};

export const SecurityEntities = {
  ASSESSMENT: {
    _type: 'azure_security_assessment',
    _class: ['Assessment'],
    resourceName: '[RM] Security Assessment',
  },
};

export const SecurityRelationships = {
  SUBSCRIPTION_PERFORMED_ASSESSMENT: {
    _type: 'azure_subscription_performed_security_assessment',
    sourceType: SUBSCRIPTION_ENTITY_METADATA._type,
    _class: RelationshipClass.PERFORMED,
    targetType: SecurityEntities.ASSESSMENT._type,
  },
};
