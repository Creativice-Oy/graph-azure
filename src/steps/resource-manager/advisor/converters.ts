import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { AdvisorEntities } from './constants';
import { ResourceRecommendationBase } from '@azure/arm-advisor/esm/models';

function maybeConvertToNumber(maybeNumber: string | undefined) {
  if (!isNaN(parseFloat(maybeNumber || 'NaN'))) return parseFloat(maybeNumber!);
}

export function createRecommendationEntity(
  webLinker: AzureWebLinker,
  data: ResourceRecommendationBase,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: AdvisorEntities.RECOMMENDATION._type,
        _class: AdvisorEntities.RECOMMENDATION._class,
        id: data.id,
        name: data.name,
        category: data.category,
        impact: data.impact,
        impactedField: data.impactedField,
        impactedValue: data.impactedValue,
        recommendationTypeId: data.recommendationTypeId,
        shortDescriptionProblem: data.shortDescription?.problem,
        shortDescriptionSolution: data.shortDescription?.solution,
        assessmentKey: data.extendedProperties?.assessmentKey,
        score: maybeConvertToNumber(data.extendedProperties?.score),
        resourceId: data.resourceMetadata?.resourceId,
        source: data.resourceMetadata?.source,
        open: true,
        severity: data.impact,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
