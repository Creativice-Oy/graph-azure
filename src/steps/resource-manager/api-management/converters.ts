import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { ApiManagementEntities } from './constants';
import {
  ApiManagementServiceResource,
  ApiContract,
} from '@azure/arm-apimanagement/esm/models';

export function createApiManagementServiceEntity(
  webLinker: AzureWebLinker,
  data: ApiManagementServiceResource,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: ApiManagementEntities.SERVICE._type,
        _class: ApiManagementEntities.SERVICE._class,
        id: data.id,
        name: data.name,
        public: !!data.publicIPAddresses?.length,
        function: ['api-gateway'],
        category: ['application'],
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createApiManagementApiEntity(
  webLinker: AzureWebLinker,
  data: ApiContract,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: ApiManagementEntities.API._type,
        _class: ApiManagementEntities.API._class,
        id: data.id,
        name: data.name,
        address: data.path,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
