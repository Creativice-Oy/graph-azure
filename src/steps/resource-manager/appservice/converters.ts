import {
  Entity,
  createIntegrationEntity,
  StepEntityMetadata,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { AppServicePlan, Site } from '@azure/arm-appservice/esm/models';
import { AppServiceEntities } from './constants';

export function createAppEntity(
  webLinker: AzureWebLinker,
  data: Site,
  metadata: StepEntityMetadata,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: metadata._type,
        _class: metadata._class,
        id: data.id,
        name: data.name,
        type: data.type,
        kind: data.kind?.split(','),
        location: data.location,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createAppServicePlanEntity(
  webLinker: AzureWebLinker,
  data: AppServicePlan,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: AppServiceEntities.APP_SERVICE_PLAN._type,
        _class: AppServiceEntities.APP_SERVICE_PLAN._class,
        id: data.id,
        name: data.name,
        type: data.type,
        kind: data.kind?.split(','),
        location: data.location,
        'sku.name': data.sku?.name,
        'sku.tier': data.sku?.tier,
        'sku.size': data.sku?.size,
        'sku.family': data.sku?.family,
        'sku.capacity': data.sku?.capacity,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
