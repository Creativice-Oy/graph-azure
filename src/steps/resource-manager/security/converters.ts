import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { SecurityEntities } from './constants';
import {
  SecurityAssessment,
  AzureResourceDetails,
} from '@azure/arm-security/esm/models';

function findSecurityAssessmentScannedResourceId(
  data: SecurityAssessment,
): string | undefined {
  // typescript typings are wrong, these properties use capital letters
  // testing both in case this is remediated in the API in the future.
  const source =
    data.resourceDetails.source || (data.resourceDetails as any).Source;
  if (source === 'Azure') {
    return (
      (data.resourceDetails as AzureResourceDetails).id ||
      (data.resourceDetails as any).Id
    );
  }
}

export function createAssessmentEntity(
  webLinker: AzureWebLinker,
  data: SecurityAssessment,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        ...convertProperties(data.metadata, { prefix: 'metadata' }),
        ...convertProperties(data.resourceDetails, {
          prefix: 'resourceDetails',
        }),
        _key: data.id as string,
        _type: SecurityEntities.ASSESSMENT._type,
        _class: SecurityEntities.ASSESSMENT._class,
        id: data.id,
        name: data.name,
        displayName: data.displayName,
        category: data.metadata?.category
          ? JSON.stringify(data.metadata.category)
          : 'Security Assessment',
        summary: data.displayName,
        internal: true,
        type: data.type,
        statusCode: data.status.code,
        statusCause: data.status.cause,
        statusDescription: data.status.description,

        scannedResourceId: findSecurityAssessmentScannedResourceId(data),

        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
