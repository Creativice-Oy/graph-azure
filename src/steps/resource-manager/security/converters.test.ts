import { createAzureWebLinker } from '../../../azure';
import { createAssessmentEntity } from './converters';
import {
  SecurityAssessment,
  AzureResourceDetails,
} from '@azure/arm-security/esm/models';
import { SecurityEntities } from './constants';

const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createAssessmentEntity', () => {
  test('properties transferred', () => {
    const data: SecurityAssessment = {
      id:
        '/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage/providers/Microsoft.Security/assessments/51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6',
      name: '51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6',
      type: 'Microsoft.Security/assessments',
      displayName: 'Storage account public access should be disallowed',
      status: { code: 'Unhealthy' },
      // Azure SDK for JS typing defines AzureResourceDetails as { source: "Azure", readonly id?: string }, although
      // the Assessments API returns capitalized property names.
      resourceDetails: ({
        Source: 'Azure',
        Id:
          '/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage',
      } as unknown) as AzureResourceDetails,
    };

    const assessmentEntity = createAssessmentEntity(webLinker, data);

    expect(assessmentEntity).toMatchGraphObjectSchema({
      _class: SecurityEntities.ASSESSMENT._class,
    });
    expect(assessmentEntity).toMatchSnapshot();
  });
});
