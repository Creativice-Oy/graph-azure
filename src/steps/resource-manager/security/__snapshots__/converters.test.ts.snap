// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAssessmentEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Assessment",
  ],
  "_key": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage/providers/Microsoft.Security/assessments/51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "displayName": "Storage account public access should be disallowed",
        "id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage/providers/Microsoft.Security/assessments/51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6",
        "name": "51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6",
        "resourceDetails": Object {
          "Id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage",
          "Source": "Azure",
        },
        "status": Object {
          "code": "Unhealthy",
        },
        "type": "Microsoft.Security/assessments",
      },
    },
  ],
  "_type": "azure_security_assessment",
  "active": undefined,
  "category": "Security Assessment",
  "createdOn": undefined,
  "displayName": "Storage account public access should be disallowed",
  "id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage/providers/Microsoft.Security/assessments/51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6",
  "internal": true,
  "name": "51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6",
  "resourceDetailsId": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage",
  "resourceDetailsSource": "Azure",
  "scannedResourceId": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage",
  "statusCause": undefined,
  "statusCode": "Unhealthy",
  "statusDescription": undefined,
  "summary": "Storage account public access should be disallowed",
  "type": "Microsoft.Security/assessments",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/microsoft.storage/storageaccounts/ndowmon1j1devblobstorage/providers/Microsoft.Security/assessments/51fd8bb1-0db4-bbf1-7e2b-cfcba7eb66a6",
}
`;
